node/narwhal/src/helpers/codec.rs

```diff
@@ -84,3 +92,244 @@ impl<N: Network> Decoder for EventCodec<N> {
+                let mut bytes = BytesMut::new();
+                match message_or_bytes {
+                    // Don't allow sending raw bytes after the noise handshake has completed.
+                    EventOrBytes::Bytes(_) => panic!("Unsupported post-handshake"),
+                    EventOrBytes::Event(event) => self.event_codec.encode(event, &mut bytes)?,
+                }
+
+                // Chunk the payload if necessary and encrypt with Noise.
+                //
+                // A Noise transport message is simply an AEAD ciphertext that is less than or
+                // equal to 65535 bytes in length, and that consists of an encrypted payload plus
+                // 16 bytes of authentication data.
+                //
+                // See: https://noiseprotocol.org/noise.html#the-handshakestate-object
+                let encrypted_chunks = bytes
+                    .par_chunks(MAX_MESSAGE_LEN - 16)
+                    .enumerate()
+                    .map(|(nonce_offset, plaintext_chunk)| {
+                        let mut buffer = vec![0u8; MAX_MESSAGE_LEN];
+                        let len = noise
+                            .state
+                            .write_message(noise.tx_nonce + nonce_offset as u64, plaintext_chunk, &mut buffer)
+                            .map_err(|e| Self::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+                        buffer.truncate(len);
+
+                        Ok(buffer)
+                    })
+                    .collect::<io::Result<Vec<Vec<u8>>>>()?;
+
+                let mut buffer = BytesMut::with_capacity(encrypted_chunks.len());
+                for chunk in encrypted_chunks {
+                    buffer.extend_from_slice(&chunk);
+                    noise.tx_nonce += 1;
+                }
+
+                buffer
+            }
+        };
+
+        // Encode the resulting ciphertext using the length-delimited codec.
+        self.codec.encode(ciphertext.freeze(), dst)
+    }
+}
+
+impl<N: Network> Decoder for NoiseCodec<N> {
+    type Error = io::Error;
+    type Item = EventOrBytes<N>;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        // Decode the ciphertext with the length-delimited codec.
+        let Some(bytes) = self.codec.decode(src)? else {
+            return Ok(None);
+        };
+
+        let msg = match self.noise_state {
+            NoiseState::Handshake(ref mut noise) => {
+                // Decrypt the ciphertext in handshake mode.
+                let mut buffer = [0u8; MAX_MESSAGE_LEN];
+                let len = noise.read_message(&bytes, &mut buffer).map_err(|_| io::ErrorKind::InvalidData)?;
+
+                Some(EventOrBytes::Bytes(Bytes::copy_from_slice(&buffer[..len])))
+            }
+
+            NoiseState::PostHandshake(ref mut noise) => {
+                // Noise decryption.
+                let decrypted_chunks: Vec<io::Result<Vec<u8>>> = bytes
```

Comment:
```suggestion
                let decrypted_chunks: io::Result<Vec<Vec<u8>>> = bytes
```
(also needs an adjustment when chunking)